temporalio/client.py

```diff
@@ -0,0 +1,935 @@
+        ]
+        if input.start_signal is not None:
+            req = temporalio.api.workflowservice.v1.SignalWithStartWorkflowExecutionRequest(
+                signal_name=input.start_signal
+            )
+            if input.start_signal_args:
+                req.signal_input.payloads.extend(
+                    await self._client.data_converter.encode(input.start_signal_args)
+                )
+        else:
+            req = temporalio.api.workflowservice.v1.StartWorkflowExecutionRequest()
+        req.namespace = self._client.namespace
+        req.workflow_id = input.id
+        req.workflow_type.name = input.workflow
+        req.task_queue.name = input.task_queue
+        if input.args:
+            req.input.payloads.extend(
+                await self._client.data_converter.encode(input.args)
+            )
+        if input.execution_timeout is not None:
+            req.workflow_execution_timeout.FromTimedelta(input.execution_timeout)
+        if input.run_timeout is not None:
+            req.workflow_run_timeout.FromTimedelta(input.run_timeout)
+        if input.task_timeout is not None:
+            req.workflow_task_timeout.FromTimedelta(input.task_timeout)
+        req.identity = self._client.identity
+        req.request_id = str(uuid.uuid4())
+        req.workflow_id_reuse_policy = cast(
+            "temporalio.api.enums.v1.WorkflowIdReusePolicy.ValueType",
+            int(input.id_reuse_policy),
+        )
+        if input.retry_policy is not None:
+            input.retry_policy.apply_to_proto(req.retry_policy)
+        req.cron_schedule = input.cron_schedule
+        if input.memo is not None:
+            for k, v in input.memo.items():
+                req.memo.fields[k] = (await self._client.data_converter.encode([v]))[0]
+        if input.search_attributes is not None:
+            for k, v in input.search_attributes.items():
+                req.search_attributes.indexed_fields[k] = (
+                    await self._client.data_converter.encode([v])
+                )[0]
+        if input.header is not None:
+            for k, v in input.header.items():
+                req.header.fields[k] = (await self._client.data_converter.encode([v]))[
+                    0
+                ]
+
+        # Start with signal or just normal start
+        resp: Union[
+            temporalio.api.workflowservice.v1.SignalWithStartWorkflowExecutionResponse,
+            temporalio.api.workflowservice.v1.StartWorkflowExecutionResponse,
+        ]
+        if isinstance(
+            req,
+            temporalio.api.workflowservice.v1.SignalWithStartWorkflowExecutionRequest,
+        ):
+            resp = await self._client.service.signal_with_start_workflow_execution(
+                req, retry=True
+            )
+        else:
+            resp = await self._client.service.start_workflow_execution(req, retry=True)
+        return WorkflowHandle(self._client, req.workflow_id, run_id=resp.run_id)
```

Comment:
When using `signal_with_start` the run ID returned might not be the `first_execution_run_id`.
I'd force the handle to use the first run ID wherever possible.

Reply:
I think it's much simpler for the user if the handle returned from start / signal_with_start always use only the workflow ID (and ideally the first_execution_run_id) and handles created with `get_workflow_handle` use whatever is provided.
In the PR I just merged for TS, there's an internal param on the handle specifying which run_id to start from for polling for result so the handle doesn't accidentally get the result of a different chain.

Reply:
I think that PR made things worse.

Someone's gonna be real upset that when they called `terminate` of their started handle, you ended up terminating a completely different workflow called from some completely different client that happens to have the same ID.

Reply:
Against what I think is better judgement, the team has decided to allow signal, query, cancel, and terminate etc to occur on workflows created in other clients :-(